plugins/spectate/spectate.cpp

```diff
@@ -0,0 +1,382 @@
+
+    if (!config.isValid()) {
+        config = World::AddPersistentData(CONFIG_KEY);
+        saveConfig();
+    } else {
+        unpause_enabled = config.ival(UNPAUSE);
+        disengage_enabled = config.ival(DISENGAGE);
+        focus_jobs_enabled = config.ival(JOB_FOCUS);
+        tick_threshold = config.ival(TICK_THRESHOLD);
+        pause_lock->unlock();
+        enable_auto_unpause(out, unpause_enabled);
+    }
+    return DFHack::CR_OK;
+}
+
+DFhackCExport command_result plugin_enable(color_ostream &out, bool enable) {
+    namespace EM = EventManager;
+    if (enable && !enabled) {
+        out.print("Spectate mode enabled!\n");
+        using namespace EM::EventType;
+        EM::EventHandler ticking(onTick, 15);
+        EM::EventHandler start(onJobStart, 0);
+        EM::EventHandler complete(onJobCompletion, 0);
+        EM::registerListener(EventType::TICK, ticking, plugin_self);
+        EM::registerListener(EventType::JOB_STARTED, start, plugin_self);
+        EM::registerListener(EventType::JOB_COMPLETED, complete, plugin_self);
+        enabled = true; // enable_auto_unpause won't do anything without this set now
+        enable_auto_unpause(out, unpause_enabled);
+    } else if (!enable && enabled) {
+        // warp 8, engage!
+        out.print("Spectate mode disabled!\n");
+        EM::unregisterAll(plugin_self);
+        // we need to retain whether auto-unpause is enabled, but we also need to disable its effect
+        bool temp = unpause_enabled;
+        enable_auto_unpause(out, false);
+        unpause_enabled = temp;
+        job_tracker.clear();
+        freq.clear();
+    }
+    enabled = enable;
+    return DFHack::CR_OK;
+}
+
+DFhackCExport command_result plugin_onstatechange(color_ostream &out, state_change_event event) {
+    if (enabled) {
+        switch (event) {
+            case SC_MAP_UNLOADED:
+            case SC_BEGIN_UNLOAD:
+            case SC_WORLD_UNLOADED:
+                our_dorf = nullptr;
+                job_watched = nullptr;
+                following_dwarf = false;
+            default:
+                break;
+        }
+    }
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_onupdate(color_ostream &out) {
+    // keeps announcement pause settings locked
+    World::Update(); // from pause.h
+    if (lock_collision) {
+        if (unpause_enabled) {
+            // player asked for auto-unpause enabled
+            World::SaveAnnouncementSettings();
+            if (World::DisableAnnouncementPausing()) {
+                // now that we've got what we want, we can lock it down
+                lock_collision = false;
+            }
+        } else {
+            if (World::RestoreAnnouncementSettings()) {
+                lock_collision = false;
+            }
+        }
+    }
+    while (unpause_enabled && !world->status.popups.empty()) {
+        // dismiss announcement popup(s)
+        Gui::getCurViewscreen(true)->feed_key(interface_key::CLOSE_MEGA_ANNOUNCEMENT);
```

Comment:
are you certain this is the right interface key for *all* popups? Does this need a failsafe that bails out of the loop if `world->status.popups.size()` does not decrement?

Reply:
No. This is code pilfered from df-ai. I will double check, but I think that's all it does.

Reply:
Yes it is. So the presumption here, is @BenLubar has functional code in df-ai
https://github.com/BenLubar/df-ai/blob/345855e39ecdb0e4614620cad12fcee01ac57f1c/pause.cpp#L25-L36

Reply:
Ben hasn't replied, and even if this is the right interface key to send today, it might not be true in the future. Avoiding an infinite loop would be very simple. I think we should add the ~2 lines of code that would prevent it.

Reply:
They have been added, see the thread I pinged you in